### DISCUSSION TITLE
Send CTRL-W (h|j|k|l)

### DISCUSSION THREAD

**Discussion by wilriker** [wezterm/wezterm#995] [Q&A]
I am trying to migrate from `tmux` to `wezterm` (I hardly ever use the tmux-server-survives-window-close and use it for multiplexing 99.9% of the time only). I have configured `SHIFT+ARROWKEY` to move between panes in `tmux` and I can replicate that behavior in `wezterm`.
But I also have a little extended configuration in `tmux` to check whether the active pane hosts a `neovim` instance and if that's the case instead of switching the pane it should rather send `CTRL-w` plus any of `h, j, k, l` to the `neovim` instance. That will then (by using a plugin) decide whether it will move around between `neovim` splits or if at the edge move out to the `tmux` pane in that direction (if any).

Now I struggle to get `wezterm` to sending `CTRL-w k` (or any other of the four characters).

My `wezterm.lua` has the following defined (removed some irrelevant settings for other keys and fonts) but now I'm stuck and help would be very much appreciated
```lua
local wezterm = require 'wezterm';

local move_around = function(window, pane, direction_wez, direction_nvim)
	if pane:get_title():sub(-4) == "NVIM" then
		window:perform_action(wezterm.action{SendString="\x17"..direction_nvim}, pane)
	else
		window:perform_action(wezterm.action{ActivatePaneDirection=direction_wez}, pane)
	end
end

wezterm.on("move-left", function(window, pane)
	move_around(window, pane, "Left", "h")
end)

wezterm.on("move-right", function(window, pane)
	move_around(window, pane, "Right", "l")
end)

wezterm.on("move-up", function(window, pane)
	move_around(window, pane, "Up", "k")
end)

wezterm.on("move-down", function(window, pane)
	move_around(window, pane, "Down", "j")
end)


return {
	disable_default_key_bindings = true,
	keys = {
		{ key = "LeftArrow",	mods = "SHIFT",	action = wezterm.action{EmitEvent="move-left"}},
		{ key = "RightArrow",	mods = "SHIFT",	action = wezterm.action{EmitEvent="move-right"}},
		{ key = "UpArrow",	mods = "SHIFT",	action = wezterm.action{EmitEvent="move-up"}},
		{ key = "DownArrow",	mods = "SHIFT",	action = wezterm.action{EmitEvent="move-down"}},
	},
}
```

Once that is fixed I also need to be able to have `neovim` send a command to `wezterm` to switch to an adjacent pane. `wezterm cli` seems to be too limited to do so but can I send it to the socket directly and if so how would be the format?

Looking forward to all answers.

Manuel

**Comment by wez** [comment]
I tried your config but modified to look for `vi` at the start of the pane title (because I use 'vim' but via a `vi` alias and that is at the start of my pane title), and shift-arrow presses do result in moving through the panes in vim:

```lua
if pane:get_title():sub(1,2) == "vi" then
```

so the `SendString` stuff appears to be working correctly.  I'd suggest setting `debug_key_events = true` in the config to help debug what's being pressed, and also do double check that the pane title has the contents you think it does by using `wezterm.log_info(pane:get_title())` to log it.

Both of those logs go to stderr, but in nightly builds you can assign a key to https://wezfurlong.org/wezterm/config/lua/keyassignment/ShowDebugOverlay.html to activate to see the recent events.

Regarding calling back into wezterm: the socket protocol is not intended to be used directly, but I'm very open to suggestions on what we need to add support for in the `wezterm cli` subcommand round out the functionality that you're looking for.

**Reply by wilriker** [reply]
I tried it with nightly build and there moving inside `nvim` works with no further config changes but I encountered something weird:
when moving between `wezterm` panes or moving from a `wezterm` pane into `nvim` (but not moving around inside `nvim`) it will visually appear like the focus has not moved (i.e. the active cursor is displayed in the pane just left and the new pane keeps the inactive cursor). But as soon as I press another key it will update the focus to whichever pane I was navigating to. So it actually has changed the pane but the visual indication is lacking behind one keypress.

Re `wezterm cli`: I personally need to use `ActivatePaneDirection`. The vim plugin I use to navigate from inside `nvim` back to `tmux` also has an option to prevent movement if the current pane (hosting `nvim`) is zoomed. I personally don't use that function but if it were accessible via `wezterm cli` I could replicate the full behavior of that plugin.
I just tried it and found that when a pane is zoomed it already prevents moving to another pane instead of unzooming and moving, so I'd like to ask for such an option (or pointing me to the appropriate config key).

**Reply by zegervdv** [reply]
To be able to replace tmux for me the `wezterm cli` would need to implement the equivalents of these `tmux` commands:
https://github.com/numToStr/Navigator.nvim/blob/0943031f2e30a0ff84212c8848f03ddb4f60554d/lua/Navigator/tmux.lua#L35-L45
So basically a command to jump to the first pane on the left/right/top/bottom and a command to query whether the current pane is zoomed.

I guess the specific plugin could then be extended to support both tmux and wezterm.

**Comment by mhanberg** [comment]
I am interested in migrating from alacritty/tmux to wezterm, and am wondering if this required any vim configuration to work?

I currently use [vim-tmux-navigator](https://github.com/christoomey/vim-tmux-navigator) and it requires both vim (through the plugin, obviously) config and tmux config to work.

**Reply by mhanberg** [reply]
It looks like `pane:get_title()` isn't going to be super robust, as I think neovim seems to have a bug with restoring the pane title when it exits.

If a `pane_tty` were to be available (similar to how it is in tmux), then i think my current config couple be wholesale ported to wezterm.

My [current config](https://github.com/mhanberg/.dotfiles/blob/main/tmux.conf#L22..L28), for context

```tmux
is_vim="ps -o state= -o comm= -t '#{pane_tty}' \
    | grep -iqE '^[^TXZ ]+ +(\\S+\\/)?g?(view|n?vim?x?)(diff)?$'"
bind-key -n C-h if-shell "$is_vim" "send-keys C-h"  "select-pane -L"
bind-key -n C-j if-shell "$is_vim" "send-keys C-j"  "select-pane -D"
bind-key -n C-k if-shell "$is_vim" "send-keys C-k"  "select-pane -U"
bind-key -n C-l if-shell "$is_vim" "send-keys C-l"  "select-pane -R"
bind-key -n C-'\' if-shell "$is_vim" "send-keys C-\\" "select-pane -l"
```

Also, i'd like to note that the config being in lua has been a delight even in my brief exploration of wezterm. Nice job!

**Reply by wez** [reply]
Exposing the tty device path is probably doable, with caveats: wezterm supports remote multiplexer panes, so there may not be a tty device path, and on Windows, there isn't an equivalent concept; the ptys are all anonymous.

wezterm has some ability to enumerate the list of processes associated with a pane to support [exit_behavior](https://wezfurlong.org/wezterm/config/lua/config/exit_behavior.html), and it should also be possible to expose that information to lua as well, so it should be possible for you to write an `is_vim` function in lua that matches against just that list of processes.

**Reply by mhanberg** [reply]
Nice!

Wezterm looks cool, thanks for creating and sharing with the world!

**Reply by aca** [reply]
I'm also trying to replace vim-tmux-navigator.
Code @wilriker provides partial solutions but for this plugin to fully work, there should be a api(cli maybe) to move pane `wezterm.action{ActivatePaneDirection="Left"}` inside vim.

Currently I couldn't find one. Is there any api for it? @wez

**Reply by wez** [reply]
There isn't an API for that at the moment.  https://github.com/wez/wezterm/issues/225 and https://github.com/wez/wezterm/issues/674 discuss some of the thinking around the possible direction, but there isn't a clear path forwards yet!

**Reply by wez** [reply]
FWIW, nightly wezterm now has [pane:get_foreground_process_name()](https://wezfurlong.org/wezterm/config/lua/pane/get_foreground_process_name.html) which works for locally spawned processes, which may make things a bit easier than searching inside the title string.

**Comment by aca** [comment]
@mhanberg @wilriker 

I created https://github.com/aca/wezterm.nvim
For now, I'm using this as a replacement.

**Reply by wilriker** [reply]
I'll try next week. But for now I _think_ the `not` in the `if` for fish shell is wrong.

**Reply by jacobfoard** [reply]
Just set this up and its working great, thanks so much @aca!

**Reply by drover25** [reply]
Thanks @aca for this! For some reason though, it doesn't work on OSX if you open wezterm by clicking the icon. I've tried to debug it, but to get logs I need to launch wezterm from a terminal, and if I launch it from a terminal it works fine :(. No idea what the difference is...

**Reply by drover25** [reply]
Just in case anyone else stumbles upon this, I fixed my issue. The `wezterm.nvim.navigator` was on the path in my shell, but not the default path. Specifying the full path of the binary resolved the issue for me. Moving the binary to somewhere on my default path would have worked too.

```
	local result = os.execute(
		"env NVIM_LISTEN_ADDRESS=/tmp/nvim"
			.. pane:pane_id()
			.. " /Users/coldis/go/bin/wezterm.nvim.navigator "
			.. direction_nvim
	)
```

**Reply by wimstefan** [reply]
Be aware that NVIM_LISTEN_ADDRESS is deprecated -> https://github.com/neovim/neovim/issues/14090#issuecomment-1116142242
Not sure if that is relevant in your case though ...

**Reply by lkhphuc** [reply]
Has anyone got this running on a remote machine, using `wezterm connect ssh ...`?
It works fine if I run it locally but does not work in on remote machine. I installed wezterm.nvim.navigator on the remote machine too.
My guess is that since I wezterm UI is running locally, the local navigator does not have control over the remote term?

**Comment by ed7ed** [comment]
Dedicate to emacs users who use `emacsclient` and Ctrl-{h,j,k,l} windows/panes navigation, similar to vim-tmux-navigator or evil-tmux-navigator:
https://github.com/illia-danko/evil-wezterm-navigator
